File: widgetssdk/src/main/java/com/glia/widgets/view/header/AppBarView.java
package com.glia.widgets.view.header;

import android.content.Context;
import android.content.res.ColorStateList;
import android.content.res.TypedArray;
import android.graphics.Typeface;
import android.util.AttributeSet;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.core.content.res.ResourcesCompat;
import androidx.core.graphics.drawable.DrawableCompat;

import com.glia.widgets.R;
import com.glia.widgets.UiTheme;
import com.glia.widgets.helper.Utils;
import com.glia.widgets.view.header.button.GliaEndButton;
import com.google.android.material.appbar.AppBarLayout;
import com.google.android.material.appbar.MaterialToolbar;

public class AppBarView extends AppBarLayout {
    private final GliaEndButton gliaEndButton;
    private final MaterialToolbar materialToolbar;
    private final TextView titleView;
    private UiTheme theme;

    public AppBarView(@NonNull Context context) {
        this(context, null);
    }

    public AppBarView(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, R.attr.gliaChatStyle);
    }

    public AppBarView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        View view = View.inflate(context, R.layout.app_bar, this);

        materialToolbar = view.findViewById(R.id.toolbar);
        titleView = view.findViewById(R.id.title);
        gliaEndButton = view.findViewById(R.id.end_button);

        materialToolbar.setNavigationContentDescription(context.getString(R.string.glia_top_app_bar_navigate_up_content_description));

        TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.AppBarView);
        Integer backIconResId = Utils.getTypedArrayIntegerValue(
                typedArray,
                context,
                R.styleable.AppBarView_backIcon,
                R.attr.gliaIconAppBarBack);
        materialToolbar.setNavigationIcon(backIconResId);
        setDefaultTitleTintColor(typedArray, context);
        setDefaultHomeButtonTintColor(typedArray, context);
        setDefaultBackgroundTintColor(typedArray, context);

        String title = Utils.getTypedArrayStringValue(typedArray, R.styleable.AppBarView_titleText);
        if (title != null) {
            titleView.setText(title);
        }

        TypedArray typedArrayView = context.obtainStyledAttributes(attrs, R.styleable.GliaView);
        this.theme = Utils.getThemeFromTypedArray(typedArrayView, this.getContext());
        setTypeface();
        DrawableCompat.setTint(materialToolbar.getMenu().findItem(R.id.leave_queue_button).getIcon(),
                ResourcesCompat.getColor(
                        getResources(),
                        Utils.getTypedArrayIntegerValue(typedArrayView,
                                context,
                                R.styleable.GliaView_chatHeaderExitQueueButtonTintColor,
                                R.attr.gliaChatHeaderExitQueueButtonTintColor
                        ),
                        this.getContext().getTheme())
        );
    }

    private void setDefaultBackgroundTintColor(TypedArray typedArray, Context context) {
        ColorStateList backgroundTintList = ContextCompat.getColorStateList(
                context,
                Utils.getTypedArrayIntegerValue(typedArray,
                        context,
                        R.styleable.AppBarView_android_backgroundTint,
                        R.attr.gliaBrandPrimaryColor
                )
        );
        materialToolbar.setBackgroundTintList(backgroundTintList);
    }

    private void setDefaultHomeButtonTintColor(TypedArray typedArray, Context context) {
        int homeButtonTintColor = ContextCompat.getColor(
                context,
                Utils.getTypedArrayIntegerValue(typedArray,
                        context,
                        R.styleable.AppBarView_lightTint,
                        R.attr.gliaChatHeaderHomeButtonTintColor
                )
        );
        materialToolbar.getNavigationIcon().setTint(homeButtonTintColor);
    }

    private void setDefaultTitleTintColor(TypedArray typedArray, Context context) {
        int titleTintColor = ContextCompat.getColor(
                context,
                Utils.getTypedArrayIntegerValue(typedArray,
                        context,
                        R.styleable.AppBarView_lightTint,
                        R.attr.gliaChatHeaderTitleTintColor
                )
        );
        titleView.setTextColor(titleTintColor);
    }


    private void setTypeface() {
        if (this.theme.getFontRes() != null) {
            Typeface fontFamily = ResourcesCompat.getFont(
                    this.getContext(),
                    this.theme.getFontRes()
            );
            titleView.setTypeface(fontFamily);
        }
    }

    public void setTheme(UiTheme uiTheme) {
        if (uiTheme == null) return;
        this.theme = Utils.getFullHybridTheme(uiTheme, this.theme);

        // icons
        materialToolbar.setNavigationIcon(theme.getIconAppBarBack());
        materialToolbar.getMenu().findItem(R.id.leave_queue_button).setIcon(theme.getIconLeaveQueue());

        // colors
        materialToolbar.setBackgroundTintList(
                ContextCompat.getColorStateList(
                        this.getContext(),
                        theme.getBrandPrimaryColor()));

        DrawableCompat.setTint(materialToolbar.getMenu().findItem(R.id.leave_queue_button).getIcon(),
                ResourcesCompat.getColor(
                        getResources(),
                        theme.getBaseLightColor(),
                        this.getContext().getTheme()));

        titleView.setTextColor(ContextCompat.getColorStateList(getContext(), theme.getGliaChatHeaderTitleTintColor()));
        materialToolbar.getNavigationIcon().setTint(ContextCompat.getColor(this.getContext(), theme.getGliaChatHeaderHomeButtonTintColor()));
        gliaEndButton.setTheme(theme);
        DrawableCompat.setTint(
                materialToolbar.getMenu().findItem(R.id.leave_queue_button).getIcon(),
                ContextCompat.getColor(getContext(), theme.getGliaChatHeaderExitQueueButtonTintColor())
        );
        setTypeface();
    }

    public void setTitle(String title) {
        titleView.setText(title);
    }

    public void showToolbar() {
        setVisibility(VISIBLE);
    }

    public void showXButton() {
        gliaEndButton.setVisibility(GONE);
        materialToolbar.getMenu().findItem(R.id.leave_queue_button).setVisible(true);
    }

    public void showEndButton() {
        gliaEndButton.setVisibility(VISIBLE);
        materialToolbar.getMenu().findItem(R.id.leave_queue_button).setVisible(false);
    }

    public void setOnBackClickedListener(OnBackClickedListener onBackClickedListener) {
        materialToolbar.setNavigationOnClickListener(view -> onBackClickedListener.onBackClicked());
    }

    public void setOnXClickedListener(OnXClickedListener onXClickedListener) {
        materialToolbar.setOnMenuItemClickListener(item -> {
            onXClickedListener.onXClicked();
            return true;
        });
    }

    public void setOnEndChatClickedListener(OnEndChatClickedListener onEndChatClickedListener) {
        gliaEndButton.setOnClickListener(v -> onEndChatClickedListener.onEnd());
    }

    public void hideLeaveButtons() {
        gliaEndButton.setVisibility(GONE);
        materialToolbar.getMenu().findItem(R.id.leave_queue_button).setVisible(false);
    }

    private MenuItem saveItem;
    private MenuItem shareItem;

    public void setMenuImagePreview() {
        materialToolbar.inflateMenu(R.menu.menu_file_preview);
        Menu menu = materialToolbar.getMenu();
        saveItem = menu.findItem(R.id.save_item);
        shareItem = menu.findItem(R.id.share_item);
    }

    public void setImagePreviewButtonListener(OnImagePreviewMenuListener listener) {
        materialToolbar.setOnMenuItemClickListener(item -> {
            int itemId = item.getItemId();
            if (itemId == R.id.share_item) {
                listener.onShareClicked();
            } else if (itemId == R.id.save_item) {
                listener.onDownloadClicked();
            }
            return true;
        });
    }

    public void setImagePreviewButtonsVisible(boolean saveItem, boolean shareItem) {
        this.saveItem.setVisible(saveItem);
        this.shareItem.setVisible(shareItem);
    }

    public interface OnImagePreviewMenuListener {
        void onShareClicked();

        void onDownloadClicked();
    }

    public interface OnBackClickedListener {
        void onBackClicked();
    }

    public interface OnXClickedListener {
        void onXClicked();
    }

    public interface OnEndChatClickedListener {
        void onEnd();
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/view/header/AppBarView.kt
package com.glia.widgets.view.header

import android.content.Context
import android.util.AttributeSet
import android.view.MenuItem
import android.widget.TextView
import androidx.annotation.DrawableRes
import androidx.core.content.withStyledAttributes
import androidx.core.view.children
import androidx.core.view.isGone
import androidx.core.view.isVisible
import androidx.core.widget.TextViewCompat
import com.glia.widgets.R
import com.glia.widgets.databinding.AppBarBinding
import com.glia.widgets.di.Dependencies
import com.glia.widgets.helper.ResourceProvider
import com.glia.widgets.helper.applyGliaThemeFont
import com.glia.widgets.helper.getColorCompat
import com.glia.widgets.helper.gliaAttrResourceId
import com.glia.widgets.helper.layoutInflater
import com.glia.widgets.helper.setLocaleContentDescription
import com.glia.widgets.helper.setLocaleNavigationContentDescription
import com.glia.widgets.helper.setLocaleText
import com.glia.widgets.helper.setText
import com.glia.widgets.helper.setTintCompat
import com.glia.widgets.locale.LocaleProvider
import com.glia.widgets.locale.LocaleString
import com.glia.widgets.view.unifiedui.applyButtonTheme
import com.glia.widgets.view.unifiedui.applyColorTheme
import com.glia.widgets.view.unifiedui.applyTextTheme
import com.glia.widgets.view.unifiedui.theme.base.HeaderTheme
import com.glia.widgets.view.unifiedui.theme.base.TextTheme
import com.google.android.material.appbar.AppBarLayout

internal class AppBarView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : AppBarLayout(context, attrs, defStyleAttr) {
    private val binding: AppBarBinding by lazy { AppBarBinding.inflate(layoutInflater, this) }

    private val toolbarTitleText: TextView?
        get() = binding.toolbar.children.firstOrNull { it.isVisible && it is TextView } as? TextView

    @DrawableRes
    private var iconAppBarBackRes: Int? = null
    private val localeProvider by lazy {
        if (isInEditMode) {
            LocaleProvider(ResourceProvider(context))
        } else {
            Dependencies.localeProvider
        }
    }

    init {
        setDefaults(attrs)
    }

    private val leaveQueueIcon: MenuItem
        get() = binding.toolbar.menu.findItem(R.id.leave_queue_button)

    private fun setDefaults(attrs: AttributeSet?) {
        iconAppBarBackRes = context.gliaAttrResourceId(
            R.attr.gliaIconAppBarBack,
            R.drawable.ic_baseline_arrow_back
        ).also {
            binding.toolbar.setNavigationIcon(it)
        }
        binding.endButton.setLocaleText(R.string.general_end)
        binding.endButton.setLocaleContentDescription(R.string.android_app_bar_end_engagement_accessibility_label)

        val titleColorRes = context.gliaAttrResourceId(
            R.attr.gliaChatHeaderTitleTintColor,
            R.color.glia_light_color
        )
        binding.title.setTextColor(getColorCompat(titleColorRes))

        val homeButtonTintColor = context.gliaAttrResourceId(
            R.attr.gliaChatHeaderHomeButtonTintColor,
            R.color.glia_light_color
        )
        binding.toolbar.navigationIcon?.setTint(getColorCompat(homeButtonTintColor))
        binding.toolbar.setLocaleNavigationContentDescription(R.string.android_app_bar_nav_up_accessibility)

        val backgroundColor = context.gliaAttrResourceId(
            R.attr.gliaBrandPrimaryColor,
            R.color.glia_primary_color
        )
        setBackgroundColor(getColorCompat(backgroundColor))

        // `titleText` is only ever set as `tools:titleText`, so this read exists purely so the four
        // layouts embedding an app bar render a title in the layout editor.
        context.withStyledAttributes(attrs, R.styleable.AppBarView) {
            getString(R.styleable.AppBarView_titleText)?.also {
                binding.title.text = it
            }
        }

        val leaveIconTint = context.gliaAttrResourceId(
            R.attr.gliaChatHeaderExitQueueButtonTintColor,
            R.color.glia_light_color
        )
        leaveQueueIcon.icon?.setTint(getColorCompat(leaveIconTint))
        leaveQueueIcon.title = localeProvider.getString(R.string.general_close)

        // A `textAppearance` cannot express "the typeface the theme asks for", so this is the one
        // value the app bar still has to resolve in code.
        context.applyGliaThemeFont(binding.title, binding.endButton)
    }

    fun hideBackButton() {
        binding.toolbar.navigationIcon = null
    }

    fun showBackButton() {
        iconAppBarBackRes?.let { binding.toolbar.setNavigationIcon(it) }
    }

    fun setTitle(title: LocaleString?) {
        binding.title.setText(title)
    }

    fun setVisibility(visibility: Boolean = true) {
        isVisible = visibility
    }

    fun showXButton() {
        binding.endButton.isGone = true
        leaveQueueIcon.isVisible = true
    }

    fun showEndButton() {
        binding.endButton.isVisible = true
        leaveQueueIcon.isVisible = false
    }

    fun hideXAndEndButton() {
        binding.endButton.isGone = true
        leaveQueueIcon.isVisible = false
    }

    fun setOnBackClickedListener(onBackClicked: OnBackClicked) {
        binding.toolbar.setNavigationOnClickListener { onBackClicked() }
    }

    fun setOnXClickedListener(onXClicked: OnXClicked) {
        binding.toolbar.setOnMenuItemClickListener {
            onXClicked()
            true
        }
    }

    fun setOnEndClickListener(onEndClicked: OnEndClicked) {
        binding.endButton.setOnClickListener { onEndClicked() }
    }

    fun hideLeaveButtons() {
        binding.endButton.isGone = true
        leaveQueueIcon.isVisible = false
    }

    internal fun resetTheme() {
        val textAttrs = intArrayOf(com.google.android.material.R.attr.textAppearanceHeadline2)
        context.obtainStyledAttributes(textAttrs).apply {
            toolbarTitleText?.also { TextViewCompat.setTextAppearance(it, getResourceId(0, 0)) }
            recycle()
        }
    }

    internal fun applyHeaderTheme(headerTheme: HeaderTheme?) {
        headerTheme?.apply {
            applyColorTheme(background?.fill)
            backButton?.iconColor?.also { binding.toolbar.setNavigationIconTint(it.primaryColor) }
            text?.also(::applyTitleTheme)
            closeButton?.iconColor?.also { leaveQueueIcon.icon?.setTintCompat(it.primaryColor) }
            endButton?.also { binding.endButton.applyButtonTheme(it) }
        }
    }

    private fun applyTitleTheme(textTheme: TextTheme) {
        if (textTheme.textColor != null && toolbarTitleText == null) {
            binding.toolbar.setTitleTextColor(textTheme.textColor.primaryColor)
            return
        }

        toolbarTitleText?.also { it.applyTextTheme(textTheme) }
    }

    fun interface OnBackClicked {
        operator fun invoke()
    }

    fun interface OnXClicked {
        operator fun invoke()
    }

    fun interface OnEndClicked {
        operator fun invoke()
    }
}
